File: data_structures/javascript/MaxHeap.js
class MaxHeap {
    constructor() {
        this.root = null;
        //where children are all stored at 
        //parentIndex * 2 + 1 and parentIndex * 2 + 2
        this.heap = [];
    }

    //O(log(n)) where n is number of nodes in tree
    insert(value) {
        this.heap.push(value);
        this.heap.siftUp(this.heap.length - 1);
    }

    leftChild(parentIndex) {
        return parentIndex * 2 + 1;
    }

    rightChild(parentIndex) {
        return parentIndex * 2 + 2;
    }

    parentIndex(childIndex) {
        return [Math.floor((childIndex - 1) / 2)];
    }

    swap(i, j) {
        const temp = this.heap[i];
        this.heap[i] = this.heap[j];
        this.heap[j] = temp;
    }

    //O(1)
    getMax() {
        return this.heap[0];
    }

    //O(log(n)) where n is number of nodes in tree
    extractMax() {
        const max = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.siftDown();
        return max;
    }

    siftUp(index) {
        while (index > 0) {
            let parentIndex = this.parentIndex(index);
            if (this.heap[parentIndex] < this.heap[index]) {
                this.swap(parentIndex, index);
            }
            index = parent;
        }
    }

    siftDown() {
        let index = 0;
        let indexToSwap = null;
        while (true) {
            let leftChild = this.leftChild(index);
            if (this.heap[index] < this.heap[leftChild]) {
                indexToSwap = leftChild;
            }
            let rightChild = this.rightChild(index);
            if (this.heap[index] < this.heap[rightChild] && (this.heap[leftChild] === null ||
                    this.heap[leftChild] !== null && this.heap[rightChild] > this.heap[leftChild])) {
                indexToSwap = rightChild;
            }
            if (!indexToSwap) {
                break;
            } else {
                this.swap(index, indexToSwap);
                index = indexToSwap;
            }
        }
    }
}

File: data_structures/ruby/max_heap.rb
class MaxHeap
    def initialize
        @heap = []
    end

    def insert(value)
        @heap.push(value)
        bubble_up
    end

    def extract_max
        max = @heap.first 
        @heap.first = @heap.pop
        bubble_down
        return max
    end

    def bubble_up 
        current_index = @heap.length - 1
        until current_index == 0
            parent_index = (current_index + 1 / 2) - 1
            if @heap[parent_index] < @heap[current_index]
                @heap[parent_index], @heap[current_index] = @heap[current_index], @heap[parent_index]
            end
            current_index = parent_index
        end
    end

    def bubble_down
        current_index = 0
        loop do 
            child_1_index = (current_index + 1) * 2
            child_2_index = child_1_index - 1
            index_to_swap = nil

            if @heap[current_index] < @heap[child_1_index]
                index_to_swap = child_1_index
            end

            if @heap[current_index] < @heap[child_2_index] && @heap[child_1_index].nil? ||
                !@heap.child_1_index.nil? && @heap[child_1_index] < @heap[child_2_index]
                index_to_swap = child_2_index
            end

            if !index_to_swap.nil?
                @heap[index_to_swap], @heap[index] = @heap[index], @heap[index_to_swap]
                index = index_to_swap
            else 
                break
            end
        end
    end
end